### PR TITLE
[hail/ptypes] PNDArray restructure 

### PR DESCRIPTION
**Pull request by akotlar**
stacks on #7733

cc @johnc1231

### CODE REVIEW

**Comment by johnc1231** [comment]
There are going to be a few conflicts since QR PR just went in. Sorry about that. I added some new PNDArray functions and renamed some others to indicate row vs column major.

**Comment by akotlar** [comment]
No worries

**Comment by akotlar** [comment]
John, I merged in your changes. Let me know if you had something else in mind for an interface. I moved all implementations, besides codeOrdering to PCanonicalNDArray because in my mind it's harder to track implementations across multiple loci, but let me know if you feel PNDArray needs some default methods (for instance, I could imagine numElements to be this way, but decided not to, because its signature really depends on the shape signature, which lives on PCanonicalNDArray)

**Comment by johnc1231** [comment]
Looks good to me. PNDArray is a little unstable at the moment and probably going to get shaken up in January, so I don't have a clear sense of any must have default methods.

**Comment by akotlar** [comment]
ready to look at

**Comment by akotlar** [comment]
> Looks good to me. PNDArray is a little unstable at the moment and probably going to get shaken up in January, so I don't have a clear sense of any must have default methods.

cool, we'll stay on top of it then.

File: hail/src/main/scala/is/hail/expr/types/physical/PCanonicalNDArray.scala [hunk 1]
```diff
@@ -0,0 +1,211 @@
+package is.hail.expr.types.physical
+
+import is.hail.annotations.{ Region, StagedRegionValueBuilder, UnsafeOrdering}
+import is.hail.asm4s.{Code, MethodBuilder, _}
+
+final case class PCanonicalNDArray(elementType: PType, nDims: Int, required: Boolean = false) extends PNDArray  {
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/types/physical/PCanonicalNDArray.scala]: The implementations are a straight copy paste, exception val modifiers on representation and derived properties

File: hail/src/main/scala/is/hail/expr/types/physical/PNDArray.scala [hunk 2]
```diff
@@ -1,224 +1,57 @@
...
-      srvb.addBaseStruct(this.shape.pType, shapeBuilder),
-      srvb.advance(),
-      srvb.addBaseStruct(this.strides.pType, stridesBuilder),
-      srvb.advance(),
-      srvb.addIRIntermediate(this.representation.fieldType("data"))(data),
-      srvb.end()
-    ))
-  }
+    stridesBuilder: (StagedRegionValueBuilder => Code[Unit]), data: Code[Long], mb: MethodBuilder): Code[Long]
 }
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/types/physical/PNDArray.scala]: paid attention to trailing whitespace this time

File: hail/src/main/scala/is/hail/expr/types/physical/PNDArray.scala [hunk 3]
```diff
@@ -1,224 +1,57 @@
...
+
+abstract class PNDArray extends PType {
+  val elementType: PType
+  val nDims: Int
+
   lazy val virtualType: TNDArray = TNDArray(elementType.virtualType, Nat(nDims), required)
   assert(elementType.required, "elementType must be required")
 
-  def _asIdent: String = s"ndarray_of_${elementType.asIdent}"
+  override def codeOrdering(mb: EmitMethodBuilder, other: PType): CodeOrdering = throw new UnsupportedOperationException
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/types/physical/PNDArray.scala]: get rid of this?

File: hail/src/main/scala/is/hail/expr/types/physical/PNDArray.scala [hunk 4]
```diff
@@ -1,224 +1,57 @@
...
-    val targetFirstElementAddress = this.data.pType.firstElementOffset(targetAddress, (nRows * nCols).toI)
-    val currentElement = Region.loadDouble(colMajorFirstElementAddress + colMajorCoord * 8L)
-
-    Code.forLoop(rowIndex := 0L, rowIndex < nRows, rowIndex := rowIndex + 1L,
-      Code.forLoop(colIndex := 0L, colIndex < nCols, colIndex := colIndex + 1L,
-        Region.storeDouble(targetFirstElementAddress + rowMajorCoord * 8L, currentElement)
-      )
-    )
-  }
+  def codeOrdering(mb: EmitMethodBuilder, other: PType): CodeOrdering = throw new UnsupportedOperationException
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/types/physical/PNDArray.scala]: not an override, so must be implemented. Kept John's solution to this, minus the override

**Comment by akotlar** [comment]
bump, should be good to go

**Comment by johnc1231** [comment]
I'll look over this, since Tim is out today

**Pull request merged**